cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java

```diff
@@ -105,12 +113,33 @@ protected void runOneIteration() throws Exception {
     if (threadPoolExecutor != null) {
       for (EventReader<StartProgramEvent> reader : readers) {
         threadPoolExecutor.execute(() -> {
-          processEvents(reader);
+          // Only attempt to process event if there is no max or the current count is less than max
+          if (hasNominalCapacity()) {
```

Comment:
Is it worth logging here if there is no capacity? This could spam the log when the instance is very busy, so may be log once in every few message?

Reply:
Great idea. We can log.trace it to avoid being chatty.